full-stack-intro/client/src/components/Todos.tsx

```diff
@@ -0,0 +1,159 @@
+/* eslint-disable @typescript-eslint/no-unused-vars -- Remove me */
+/* eslint-disable @typescript-eslint/no-empty-function -- Remove me */
+import { useEffect, useState } from 'react';
+import { PageTitle } from './PageTitle';
+import { TodoList } from './TodoList';
+import { TodoForm } from './TodoForm';
+
+export type UnsavedTodo = {
+  task: string;
+  isCompleted: boolean;
+};
+export type Todo = UnsavedTodo & {
+  todoId: number;
+};
+
+export function Todos() {
+  const [todos, setTodos] = useState<Todo[]>([]);
+  const [isLoading, setIsLoading] = useState(true);
+  const [error, setError] = useState<unknown>();
+
+  /* Implement useEffect to fetch all todos. Hints are at the bottom of the file. */
+  useEffect(() => {
+    async function fetchTodos() {
+      try {
+        const response = await fetch('api/todos');
+        if (!response.ok) throw new Error('fetch request failed');
+        const data = (await response.json()) as Todo[];
+        setTodos(data);
+      } catch (err) {
+        console.log(err);
+        setError(err);
+      } finally {
+        setIsLoading(false);
+      }
+    }
+    fetchTodos();
+  }, []);
+
+  /* Implement addTodo to add a new todo. Hints are at the bottom of the file. */
+  async function addTodo(newTodo: UnsavedTodo) {
+    try {
+      const response = await fetch('/api/todos', {
+        method: 'POST',
+        headers: {
+          'Content-Type': 'application/json',
+        },
+        body: JSON.stringify(newTodo),
+      });
+      if (!response.ok) throw new Error('failed to post');
+      const data = (await response.json()) as Todo;
+      setTodos([...todos, data]);
+    } catch (err) {
+      console.log(err);
```

Comment:
You need to display an error message to the user somehow, or they won't know something went wrong and will likely get frustrated! For now, just `alert(err.message);` is likely good enough.